File: Java-Core/src/main/java/presto/statement/ReturnStatement.java
package presto.statement;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.runtime.VoidResult;
import presto.type.IType;
import presto.type.VoidType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class ReturnStatement extends SimpleStatement {
	
	IExpression expression;
	
	public ReturnStatement(IExpression expression) {
		this.expression = expression;
	}

	public IExpression getExpression() {
		return expression;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append("return");
		if(expression!=null) {
			writer.append(" ");
			expression.toDialect(writer);
		}
	}
	
	@Override
	public String toString() {
		return "return " + expression==null ? "" : expression.toString();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj==this)
			return true;
		if(obj==null)
			return false;
		if(!(obj instanceof ReturnStatement))
			return false;
		ReturnStatement other = (ReturnStatement)obj;
		return this.getExpression().equals(other.getExpression());
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		return expression==null ? VoidType.instance() : expression.check(context);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		if(expression==null)
			return VoidResult.instance();
		else
			return expression.interpret(context);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/statement/ReturnStatement.js
var SimpleStatement = require("./SimpleStatement").SimpleStatement;
var VoidResult = require("../runtime/VoidResult").VoidResult;
var VoidType = require("../type/VoidType").VoidType;

function ReturnStatement(expression) {
	SimpleStatement.call(this);
	this.expression = expression || null;
	return this;
}

ReturnStatement.prototype = Object.create(SimpleStatement.prototype);
ReturnStatement.prototype.constructor = ReturnStatement;

ReturnStatement.prototype.toString = function() {
	return "return " + this.expression==null ? "" : this.expression.toString();
}

ReturnStatement.prototype.toDialect = function(writer) {
    writer.append("return");
    if(this.expression!=null) {
        writer.append(" ");
        this.expression.toDialect(writer);
    }
};


ReturnStatement.prototype.equals = function(obj) {
	if(obj==this) {
		return true;
	}
	if(obj==null) {
		return false;
	}
	if(!(obj instanceof ReturnStatement)) {
		return false;
	}
    if(this.expression==obj.expression)
        return true;
    else if(this.expression==null || obj.expression==null)
        return false;
	else
        return this.expression.equals(obj.expression);
};

ReturnStatement.prototype.check = function(context) {
	return this.expression==null ? VoidType.instance : this.expression.check(context);
};

ReturnStatement.prototype.interpret= function(context) {
    if(this.expression==null)
        return VoidResult.instance;
    else
        return this.expression.interpret(context);
};

exports.ReturnStatement = ReturnStatement;
